demos/elastodynamics/elastodynamics-mwe.py

```diff
@@ -0,0 +1,239 @@
+
+# Select whether or not the mass matrix will be lumped::
+lump_mass = True
+
+# Set time parameters::
+T                = 0.4
+dt               = 1e-3
+Nsteps_numerical = int(T/dt)
+t                = 0.
+step             = 0
+
+# We define Young's modulus and Poisson's ratio::
+E  = 50.0e6
+nu = 0.3
+
+# Next we calculate Lame's constants::
+lmbda   = E*nu/((1 + nu)*(1 - 2*nu))
+mu      = E/(2*(1 + nu))
+rho     = 1986.
+
+# The loading parameters are::
+amplitude = 40.0e3
+omega     = 50.
+
+# We create a rectangular column mesh where the number of elements in the
+# :math:`z`-direction is allowed to vary to demonstrate convergence::
+height = 20.
+width  = 1.
+nx     = 1
+ny     = 1
+nz     = 20
+mesh   = BoxMesh(nx, ny, nz, width, width, height)
+
+# Set up outputs::
+u_numerical    = np.zeros((Nsteps_numerical+2))
+time_numerical = np.linspace(0, T, Nsteps_numerical+2)
+
+# We first compute all surface areas of the mesh and then assign
+# the differential area element :math: `\Gamma_N`:: at the top surface
+ds      = Measure('ds', mesh)
+Gamma_N = ds(6)
+
+# We choose a degree 1 continuous vector function space, and set up
+# the function space and functions::
+W = VectorFunctionSpace(mesh, 'Lagrange', 1)
+u = TrialFunction(W)
+w = TestFunction(W)
+
+u_n   = Function(W, name='displacement')
+u_np1 = Function(W)
+u_nm1 = Function(W)
+
+# Next we define the traction vector which will be updated in the time loop::
+traction = Constant((0., 0., 0.))
+
+# We next establish the Dirichlet boundary conditions::
+fixedSides  = Constant(0.)
+fixedBottom = as_vector((0.,0.,0.))
+
+fixedLeft_BC_x  = DirichletBC(W.sub(0), fixedSides, 1)
+fixedLeft_BC_y  = DirichletBC(W.sub(1), fixedSides, 1)
+fixedRight_BC_x = DirichletBC(W.sub(0), fixedSides, 2)
+fixedRight_BC_y = DirichletBC(W.sub(1), fixedSides, 2)
+fixedBack_BC_x  = DirichletBC(W.sub(0), fixedSides, 3)
+fixedBack_BC_y  = DirichletBC(W.sub(1), fixedSides, 3)
+fixedFront_BC_x = DirichletBC(W.sub(0), fixedSides, 4)
+fixedFront_BC_y = DirichletBC(W.sub(1), fixedSides, 4)
+fixedBottom_BC  = DirichletBC(W, fixedBottom, 5)
+
+bcSet = [fixedLeft_BC_x, fixedRight_BC_x, fixedBack_BC_x, fixedFront_BC_x, \
+         fixedLeft_BC_y, fixedRight_BC_y, fixedBack_BC_y, fixedFront_BC_y, \
+         fixedBottom_BC]
+
+print("Computing numerical solution...")
+
+if lump_mass:
+    lumped_mass = assemble((rho/(dt*dt))*dot(w, Function(W).interpolate(as_vector([1,1,1]))) * dx)
+    # Time-stepping loop::
+    while t < T + EPSILON:
+        # Update time step::
+        # print(t)
+        t    += dt
+        step += 1
+        # Update load::
+        traction.assign((0., 0., -0.5*amplitude*(1 - cos(omega*t))))
+        # Assemble right-hand side::
+        rhs = assemble(-inner(sigma(u_np1), epsilon(w))*dx + dot(traction, w)*Gamma_N)
+        us  = Function(W).assign(rhs)
+        # Divide by lumped mass matrix::
+        us /= lumped_mass
+        # Update displacements::
+        u_np1.assign(us + 2*u_n - u_nm1)
```

Comment:
OK, so in this case we've done:
```
rho/dt^2 (u, w) = - (sigma(u_{n-1}), eps(w)) + (traction, w)_F + rho/dt^2 (u_{n-1} - 2 u_n, w)
```
Which after rearranging is
```
(u, w) = dt^2/rho * [(traction, w)_F - (sigma(u_{n-1}), eps(w))] + (u_{n-1} - 2 u_n, w)
```
To mass lump, we need to divide this through by the lumped mass, call it `M_lump`, but I think you're only dividing through the first of the two terms by the lumped mass and obtaining:
```
u_{n+1} = dt^2/rho M_lump^{-1} * [(traction, w)_F - (sigma(u_{n-1}), eps(w))] + (u_{n-1} - 2 u_n, w)
```

This is correct as long as `M_lump` is `rho/dt^2 I` where `I` is the identity. But I think you're likely missing a scaling by `detJ` on each cell or similar.

I think we instead want to write the left and right hand sides:


```
t = Constant(t)
traction = as_vector([0, 0, -0.5*amplitude*(1 - cos(omega * t))])
a = inner(u, w)*dx
L = dt**2/rho * (inner(traction, w)*Gamma_N - inner(sigma(u_nm1), epsilon(w))*dx) + inner(u_nm1 - 2*un, w)*dx
# Now produce the lumped mass matrix:
alump = replace(a, {u: as_vector([1, 1, 1])})
lumped_mass = assemble(alump)
while t < T + EPSILON:
    # u_{n+1} <- M_lump^{-1} L (first by assigning the assembly of L and then dividing by the lumped mass)
    t.assign(t + dt)
    assemble(L, tensor=u_np1)
    u_np1 /= lumped_mass
```

Note that I defined the `traction` vector which is time-dependent symbolically, so when we update `t`, the assembly of the `rhs` notices. This is also the right thing to do because each assembly with your previous approach would require JIT-compilation of the form defining the RHS (because you have a literal number, `t` in the form that changes at each step).

Can you have a go at these changes to see if it fixes the mass-lumped problem?

I think if so, it will also allow us to simplify the demo a bit for the non-mass-lumped case: notice how we define the same forms we'll need for the full mass solve, only in that case we won't do this `replace` dance.